bindings/src/main/java/io/pravega/storage/hdfs/HDFSChunkStorage.java

```diff
@@ -0,0 +1,350 @@
+            // of bytes to copy.
+            IOUtils.copyBytes(data, stream, (long) length, false);
+
+            stream.flush();
+        } catch (IOException e) {
+            throw convertException(handle.getChunkName(), "doWrite", e);
+        }
+        return length;
+    }
+
+    @Override
+    public int doConcat(ConcatArgument[] chunks) throws ChunkStorageException {
+        ensureInitializedAndNotClosed();
+        int length = 0;
+        try {
+            val sources = new Path[chunks.length - 1];
+            this.fileSystem.truncate(getFilePath(chunks[0].getName()), chunks[0].getLength());
+            for (int i = 1; i < chunks.length; i++) {
+                val chunkLength = chunks[i].getLength();
+                this.fileSystem.truncate(getFilePath(chunks[i].getName()), chunkLength);
+                sources[i - 1] = getFilePath(chunks[i].getName());
+                length += chunkLength;
+            }
+            // Concat source file into target.
+            this.fileSystem.concat(getFilePath(chunks[0].getName()), sources);
+        } catch (IOException e) {
+            throw convertException(chunks[0].getName(), "doConcat", e);
+        }
+        return length;
+    }
+
+    @Override
+    protected boolean doTruncate(ChunkHandle handle, long offset) throws UnsupportedOperationException {
+        throw new UnsupportedOperationException(getClass().getName() + " does not support chunk truncation.");
+    }
+
+    @Override
+    protected void doSetReadOnly(ChunkHandle handle, boolean isReadOnly) throws ChunkStorageException {
+        try {
+            this.fileSystem.setPermission(getFilePath(handle.getChunkName()), isReadOnly ? READONLY_PERMISSION : READWRITE_PERMISSION);
+        } catch (IOException e) {
+            throw convertException(handle.getChunkName(), "doSetReadOnly", e);
+        }
+    }
+
+    //endregion
+
+    //region Storage Implementation
+
+    @SneakyThrows(IOException.class)
+    public void initialize() {
+        Exceptions.checkNotClosed(this.closed.get(), this);
+        Preconditions.checkState(this.fileSystem == null, "HDFSStorage has already been initialized.");
+        Configuration conf = new Configuration();
+        conf.set("fs.default.name", this.config.getHdfsHostURL());
+        conf.set("fs.default.fs", this.config.getHdfsHostURL());
+        conf.set("fs.hdfs.impl", "org.apache.hadoop.hdfs.DistributedFileSystem");
+
+        // We do not want FileSystem to cache clients/instances based on target URI.
+        // This allows us to close instances without affecting other clients/instances. This should not affect performance.
+        conf.set("fs.hdfs.impl.disable.cache", "true");
+        if (!this.config.isReplaceDataNodesOnFailure()) {
+            // Default is DEFAULT, so we only set this if we want it disabled.
+            conf.set("dfs.client.block.write.replace-datanode-on-failure.policy", "NEVER");
+        }
+
+        this.fileSystem = openFileSystem(conf);
```

Comment:
do we have a way to pass additional configuration parameters to this client ?
I mean, probably users would like to add additional configurations, in segment store configuration file
`hdfs.customconfig.xxxx=yyyyy `

Reply:
The purpose of this PR was not to tweak the HDFS client. We can take that challenge in a separate issue. We are setting all the configs that we believe we need, and define the "pass-through" values in our own config. Otherwise we'd have an unmanageable zoo of configurations, most of which will be beyond our control.

Reply:
@andreipaduroiu agreed.
@sachin-j-joshi probably there is not need to create a ticket if no user asks for it.